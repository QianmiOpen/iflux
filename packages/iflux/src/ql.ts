import { IQLangProps, TPath, TQLang } from './types';

let id = 0;

export class QueryLang {
  private id: number;
  private name: string;
  private lang: TQLang;

  constructor(props: IQLangProps) {
    const { name, lang } = props;
    this.id = ++id;
    this.name = name;
    this.lang = lang;
  }

  meta() {
    const lang = [...this.lang];
    const handler = lang.pop() as Function;
    const deps = lang as Array<TPath>;

    return {
      id: this.id,
      name: this.name,
      deps,
      handler
    };
  }
}

//factory method
export function QL(lang: TQLang, name: string = '') {
  return new QueryLang({ name, lang });
}
