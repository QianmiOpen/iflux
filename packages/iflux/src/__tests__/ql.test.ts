import { QL } from '../ql';

test('create ql', () => {
  const callback = (hello, province, city) => {
    return hello + province + city;
  };

  const ql = QL(
    [
      //data path
      'hello',
      ['user', 'addr', 'province'],
      'user.addr.city',
      callback
    ],
    'testQL'
  );

  const meta = ql.meta();
  expect(meta).toMatchSnapshot();
  expect(meta).toEqual({
    id: 1,
    name: 'testQL',
    deps: [
      //data path
      'hello',
      ['user', 'addr', 'province'],
      'user.addr.city'
    ],
    handler: callback
  });
});
