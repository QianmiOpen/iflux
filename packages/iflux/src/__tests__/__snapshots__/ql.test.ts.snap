// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create ql 1`] = `
Object {
  "deps": Array [
    "hello",
    Array [
      "user",
      "addr",
      "province",
    ],
    "user.addr.city",
  ],
  "handler": [Function],
  "id": 1,
  "name": "testQL",
}
`;
